cyctl/cmd/serve.go

```diff
@@ -0,0 +1,120 @@
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"github.com/spf13/cobra"
+	"log"
+	"math/rand"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"strings"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+	"k8s.io/client-go/util/homedir"
+)
+
+var Clientset *kubernetes.Clientset
+
+var serveCMD = &cobra.Command{
+	Use:   "serve -port [port]",
+	Short: "Start the Cyclops UI",
+	Long:  "Start the Cyclops UI by forwarding the Cyclops UI service's port to a local port",
+	Run: func(cmd *cobra.Command, args []string) {
+		localPort, err := cmd.Flags().GetInt("port")
+		if err != nil {
+			log.Fatal(err)
+		}
+		startPortForwarding(Clientset, localPort)
+	},
+}
+
+func startPortForwarding(clientset *kubernetes.Clientset, localPort int) {
+	var kubeconfig *string
+	if home := homedir.HomeDir(); home != "" {
+		kubeconfig = flag.String("kubeconfig", filepath.Join(home, ".kube", "config"), "(optional) absolute path to the kubeconfig file")
+	} else {
+		kubeconfig = flag.String("kubeconfig", "", "absolute path to the kubeconfig file")
+	}
+	flag.Parse()
+
+	config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	clientset, err = kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	service, err := clientset.CoreV1().Services("cyclops").Get(context.TODO(), "cyclops-ui", metav1.GetOptions{})
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	podList, err := clientset.CoreV1().Pods("cyclops").List(context.TODO(), metav1.ListOptions{
+		LabelSelector: labels.Set(service.Spec.Selector).String(),
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+	if len(podList.Items) == 0 {
+		log.Fatal("no pods found for service cyclops-ui in namespace cyclops")
+	}
+
+	rand.Seed(time.Now().UnixNano())
+	podIndex := rand.Intn(len(podList.Items))
+	podName := podList.Items[podIndex].Name
+
+	roundTripper, upgrader, err := spdy.RoundTripperFor(config)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	path := fmt.Sprintf("/api/v1/namespaces/%s/pods/%s/portforward", "cyclops", podName)
+	hostIP := strings.TrimLeft(config.Host, "htps:/")
+	serverURL := url.URL{Scheme: "https", Path: path, Host: hostIP}
+
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: roundTripper}, http.MethodPost, &serverURL)
+
+	ports := []string{fmt.Sprintf("%d:80", localPort)}
+
+	stopChan, readyChan := make(chan struct{}, 1), make(chan struct{}, 1)
+	out, errOut := new(bytes.Buffer), new(bytes.Buffer)
+
+	forwarder, err := portforward.New(dialer, ports, stopChan, readyChan, out, errOut)
```

Comment:
I think you can just provide os.Stdout and os.Stderr instead of self-created buffers

```suggestion
	forwarder, err := portforward.New(dialer, ports, stopChan, readyChan, os.Stdout, os.Stderr)
```

Reply:
Using `Stdout` and `Stderr` causes the channel to exit:
```go
        go func() {
                for range readyChan {
                }
                if os.Stdout != nil {
                        log.Fatal("stdout: ", os.Stdout)
                } else if os.Stderr != nil {
                        fmt.Println("stderr: ", os.Stderr)
                }
        }()
```
```shell
Forwarding from 127.0.0.1:3002 -> 80
Forwarding from [::1]:3002 -> 80
2024/06/17 06:34:14 stdout: &{0xc00009a0c0}
exit status 1
```

Creating buffers is needed in this case

Reply:
Removed that part